lightning-invoice/src/payment.rs

```diff
@@ -425,28 +496,94 @@ where
+					// the router excludes the payer node. In the following lines, the payer's information is
+					// hardcoded with an inflight value of 0 so that we can correctly represent the first hop
+					// in our sliding window of two.
+					let our_node_id: PublicKey = self.payer.node_id();
+					let reversed_hops_with_payer = core::iter::once((0u64, our_node_id)).chain(
+						hops.split_last().unwrap().1.iter().map(|hop| (hop.fee_msat, hop.pubkey))).rev();
+					let mut cumulative_msat = hops.last().unwrap().fee_msat;
+
+					// Taking the reversed vector from above, we zip it with just the reversed hops list to
+					// work "backwards" of the given path, since the last hop's `fee_msat` actually represents
+					// the total amount sent.
+					for (next_hop, prev_hop) in hops.iter().rev().zip(reversed_hops_with_payer) {
+						cumulative_msat += prev_hop.0;
+						total_inflight_map
+							.entry((next_hop.short_channel_id, next_hop.pubkey < prev_hop.1))
+							.and_modify(|used_liquidity_msat| *used_liquidity_msat += cumulative_msat)
+							.or_insert(cumulative_msat);
+					}
+				}
+
+				total_inflight_map
+			} else {
+				HashMap::new()
+			}
+		};
+
+		let route = {
+			let scorer = AccountForInFlightHtlcs { scorer: &mut self.scorer.lock(), inflight_htlcs };
+
+			self.router.find_route(
+				&payer, &params, &payment_hash, Some(&first_hops.iter().collect::<Vec<_>>()),
+				&scorer
+			).map_err(|e| PaymentError::Routing(e))?
+		};
 
 		match send_payment(&route) {
-			Ok(payment_id) => Ok(payment_id),
+			Ok(payment_id) => {
+				for hops in route.paths {
+					self.process_path_inflight_htlcs(payment_hash, hops);
+				}
+				Ok(payment_id)
+			},
 			Err(e) => match e {
 				PaymentSendFailure::ParameterError(_) => Err(e),
 				PaymentSendFailure::PathParameterError(_) => Err(e),
 				PaymentSendFailure::AllFailedRetrySafe(_) => {
 					let mut payment_cache = self.payment_cache.lock().unwrap();
-					let payment_attempts = payment_cache.get_mut(&payment_hash).unwrap();
-					payment_attempts.count += 1;
-					if self.retry.is_retryable_now(payment_attempts) {
+					let payment_info = payment_cache.get_mut(&payment_hash).unwrap();
+					payment_info.attempts.count += 1;
+					if self.retry.is_retryable_now(&payment_info.attempts) {
 						core::mem::drop(payment_cache);
 						Ok(self.pay_internal(params, payment_hash, send_payment)?)
 					} else {
 						Err(e)
 					}
 				},
-				PaymentSendFailure::PartialFailure { failed_paths_retry, payment_id, .. } => {
+				PaymentSendFailure::PartialFailure { failed_paths_retry, payment_id, results } => {
+					// The ones that are `Ok()`s + `MonitorUpdateFailed` in `results` need to be used to update the map
+					// get the index of them and run to `route` to get the values we need to care about.
+					for (idx, result) in results.iter().enumerate() {
+						match result {
+							Ok(_) => {
+								if let Some(failed_path) = route.paths.get(idx) {
```

Comment:
I think we can assume the lengths of `route.paths` and `results` are the same and simply index into the former.